publish.go

```diff
@@ -0,0 +1,326 @@
+func updateRelease(configs Configs, versionCodes googleapi.Int64s, release *androidpublisher.TrackRelease) error {
+	if configs.UntrackBlockingVersions {
+		removeBlockingVersionFromRelease(release, versionCodes)
+	} else {
+		release.VersionCodes = append(release.VersionCodes, versionCodes...)
+	}
+	log.Infof("Release version codes are: %v", release.VersionCodes)
+
+	log.Printf(" assigned app versions: %v", release.VersionCodes)
+	if err := updateListing(configs, release); err != nil {
+		return fmt.Errorf("failed to update listing, reason: %v", err)
+	}
+	return nil
+}
+
+// removeBlockingVersionFromRelease removes blocking version from a given version, which would shadow the given version.
+func removeBlockingVersionFromRelease(release *androidpublisher.TrackRelease, newVersionCodes googleapi.Int64s) {
+	log.Printf("Checking app versions on release: %s", release.Name)
+	log.Infof("Current version codes: %v", release.VersionCodes)
+	log.Infof("New version codes: '%v'", newVersionCodes)
+
+	var cleanTrack bool
+	if len(release.VersionCodes) != len(newVersionCodes) {
+		log.Warnf("Mismatching app count, removing (%v) versions from release: %s", release.VersionCodes, release.Name)
+		cleanTrack = true
+	} else {
+		log.Debugf("The number of App version codes (current and new) are equal")
+		release.VersionCodes = sortAndFilterVersionCodes(release.VersionCodes, newVersionCodes)
+	}
+
+	if cleanTrack {
+		log.Infof("Clearing version codes for release %v, the new versions will be %v", release.Name, newVersionCodes)
+		release.VersionCodes = newVersionCodes
+	}
+}
+
+// sortAndFilterVersionCodes sorts and filters two set of version codes, returns the higher for each.
+func sortAndFilterVersionCodes(currentVersionCodes googleapi.Int64s, newVersionCodes googleapi.Int64s) googleapi.Int64s {
+	sort.Slice(currentVersionCodes, func(a, b int) bool { return currentVersionCodes[a] < currentVersionCodes[b] })
+	sort.Slice(newVersionCodes, func(a, b int) bool { return newVersionCodes[a] < newVersionCodes[b] })
+
+	var filteredVersionCodes googleapi.Int64s
+	for i := 0; i < len(newVersionCodes); i++ {
+		log.Debugf("Searching for shadowing versions, comparing (%v) and (%v)", currentVersionCodes[i], newVersionCodes[i])
+		if currentVersionCodes[i] < newVersionCodes[i] {
+			log.Infof("Shadowing app found, removing current (%v) version, adding new (%v)", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, newVersionCodes[i])
+		} else {
+			log.Infof("Currently released app with (%v) version is higher than new (%v), app with new version code ignored", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, currentVersionCodes[i])
+		}
+	}
+	return filteredVersionCodes
+}
+
+// releaseStatusFromConfig gets the release status from the config value of user fraction.
+func releaseStatusFromConfig(userFraction float64) string {
```

Comment:
Would merge this function with updateTracks, pass in result to GetRelease.

Reply:
Could you from other places, would not merge.